cpp/serve/engine.cc

```diff
@@ -87,7 +87,14 @@ class EngineImpl : public Engine {
     }
     LogitProcessor logit_processor =
         this->models_[0]->CreateLogitProcessor(max_num_tokens, trace_recorder);
-    Sampler sampler = this->models_[0]->CreateSampler(max_num_tokens, trace_recorder);
+    bool is_speculative = this->models_.size() > 1U;
+    // TODO: need a better sampler selection logic
+    std::string sampler_selection = "auto";
+    if (is_speculative) {
+      sampler_selection = "cpu";
+    }
+    Sampler sampler =
+        this->models_[0]->CreateSampler(max_num_tokens, sampler_selection, trace_recorder);
```

Comment:
Let's just pass the models as `const Array<Model>&` to `CreateSampler`, and select CPU sampler inside `CreateSampler` when the number of models is more than 1. The goal is to abstract, hide and not complicate the logic.